File: grimdb/sdk/swift/Tests/GrimlockerSDKTests/GrimlockerClientTests.swift
import XCTest
@testable import GrimlockerSDK

final class MockURLProtocol: URLProtocol {
    static var requestHandler: ((URLRequest) throws -> (HTTPURLResponse, Data))?

    override class func canInit(with request: URLRequest) -> Bool { true }
    override class func canonicalRequest(for request: URLRequest) -> URLRequest { request }
    override func stopLoading() {}
    override func startLoading() {
        guard let handler = MockURLProtocol.requestHandler else {
            fatalError("requestHandler not set")
        }
        do {
            let (response, data) = try handler(request)
            client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
            client?.urlProtocol(self, didLoad: data)
            client?.urlProtocolDidFinishLoading(self)
        } catch {
            client?.urlProtocol(self, didFailWithError: error)
        }
    }
}

final class GrimlockerClientTests: XCTestCase {
    var client: GrimlockerClient!

    static let baseURL = URL(string: "http://127.0.0.1:36353")!

    override func setUp() {
        super.setUp()
        let configuration = URLSessionConfiguration.ephemeral
        configuration.protocolClasses = [MockURLProtocol.self]
        client = GrimlockerClient(baseURL: Self.baseURL, token: "test-token")
    }

    override func tearDown() {
        MockURLProtocol.requestHandler = nil
        super.tearDown()
    }

    private func stubResponse(status: Int = 200, body: Data) {
        MockURLProtocol.requestHandler = { _ in
            let response = HTTPURLResponse(
                url: Self.baseURL.appendingPathComponent("api/v1"),
                statusCode: status,
                httpVersion: nil,
                headerFields: ["Content-Type": "application/json"]
            )!
            return (response, body)
        }
    }

    private func stubJSON(_ json: String, status: Int = 200) {
        stubResponse(status: status, body: json.data(using: .utf8)!)
    }

    private func encodeJSON(_ obj: Encodable) -> Data {
        try! JSONEncoder().encode(obj)
    }

    // ── Auth ──────────────────────────────────────────────────────────────────

    func testUnlockVault() async throws {
        stubJSON(#"{"success":true}"#)
        try await client.unlockVault(password: "master-password")
    }

    func testLockVault() async throws {
        stubJSON(#"{"success":true}"#)
        try await client.lockVault()
    }

    func testVaultStatus() async throws {
        stubJSON(#"{"initialized":true,"unlocked":true,"status":"ok"}"#)
        let status = try await client.vaultStatus()
        XCTAssertTrue(status.initialized)
        XCTAssertTrue(status.unlocked)
    }

    // ── Entries ───────────────────────────────────────────────────────────────

    func testListEntries() async throws {
        stubJSON(#"{"entries":[{"id":"e1","title":"Entry 1","category":"PASSWORD","created_at":1,"updated_at":2}],"total_count":1}"#)
        let entries = try await client.listEntries()
        XCTAssertEqual(entries.count, 1)
        XCTAssertEqual(entries[0].id, "e1")
    }

    func testListEntriesByCategory() async throws {
        stubJSON(#"{"entries":[]}"#)
        let entries = try await client.listEntries(category: "SSH_KEY")
        XCTAssertEqual(entries.count, 0)
    }

    func testGetEntry() async throws {
        stubJSON(#"{"id":"e1","title":"Entry","category":"PASSWORD"}"#)
        let entry = try await client.getEntry(id: "e1")
        XCTAssertEqual(entry.id, "e1")
        XCTAssertEqual(entry.title, "Entry")
    }

    func testCreateEntry() async throws {
        stubJSON(#"{"id":"new1","title":"New","category":"PASSWORD"}"#)
        let entry = try await client.createEntry(title: "New", category: "PASSWORD", fields: ["username": "alice"])
        XCTAssertEqual(entry.id, "new1")
    }

    func testUpdateEntry() async throws {
        stubJSON(#"{"success":true}"#)
        try await client.updateEntry(id: "e1", fields: ["notes": "updated"])
    }

    func testDeleteEntry() async throws {
        stubJSON(#"{"success":true}"#)
        try await client.deleteEntry(id: "e1")
    }

    func testSearchEntries() async throws {
        stubJSON(#"{"entries":[{"id":"e1","title":"GitHub","category":"PASSWORD"}],"total_count":1}"#)
        let results = try await client.searchEntries(query: "git")
        XCTAssertEqual(results.count, 1)
    }

    func testSearchEntriesWithCategory() async throws {
        stubJSON(#"{"entries":[]}"#)
        let results = try await client.searchEntries(query: "git", category: "SSH_KEY")
        XCTAssertEqual(results.count, 0)
    }

    // ── Typed helpers ─────────────────────────────────────────────────────────

    func testListPasswords() async throws {
        stubJSON(#"{"entries":[{"id":"p1","title":"GitHub","category":"PASSWORD","fields":{"username":"a","password":"b","url":"","notes":""}}],"total_count":1}"#)
        let passwords = try await client.listPasswords()
        XCTAssertEqual(passwords.count, 1)
        XCTAssertEqual(passwords[0].title, "GitHub")
    }

    func testCreatePassword() async throws {
        stubJSON(#"{"id":"p1","title":"GitHub","category":"PASSWORD"}"#)
        let p = PasswordEntry(title: "GitHub", username: "alice", password: "sec")
        let id = try await client.createPassword(p)
        XCTAssertEqual(id, "p1")
    }

    func testListSshKeys() async throws {
        stubJSON(#"{"entries":[{"id":"sk1","title":"Key","category":"SSH_KEY","fields":{"public_key":"pk","private_key":"","username":"","passphrase":"","comment":""}}],"total_count":1}"#)
        let keys = try await client.listSshKeys()
        XCTAssertEqual(keys.count, 1)
    }

    func testCreateSshKey() async throws {
        stubJSON(#"{"id":"sk1","title":"Key","category":"SSH_KEY"}"#)
        let k = SshKeyEntry(title: "Key", publicKey: "pk", privateKey: "priv", username: "")
        let id = try await client.createSshKey(k)
        XCTAssertEqual(id, "sk1")
    }

    func testListCertificates() async throws {
        stubJSON(#"{"entries":[{"id":"c1","title":"Cert","category":"CERTIFICATE","fields":{"domain":"ex.com","certificate":"crt","private_key":"key"}}],"total_count":1}"#)
        let certs = try await client.listCertificates()
        XCTAssertEqual(certs.count, 1)
    }

    func testCreateCertificate() async throws {
        stubJSON(#"{"id":"c1","title":"Cert","category":"CERTIFICATE"}"#)
        let c = CertificateEntry(title: "Cert", domain: "ex.com", certificate: "crt", privateKey: "key")
        let id = try await client.createCertificate(c)
        XCTAssertEqual(id, "c1")
    }

    // ── File Vault ────────────────────────────────────────────────────────────

    func testListFolder() async throws {
        stubJSON(#"{"folders":[{"id":"d1","name":"sub","type":"folder"}],"files":[{"id":"f1","file_name":"a.txt","mime_type":"text/plain","total_size":10,"manifest_block_id":"mb1","folder_id":""}]}"#)
        let listing = try await client.listFolder()
        XCTAssertEqual(listing.folders.count, 1)
        XCTAssertEqual(listing.files.count, 1)
    }

    func testCreateFolder() async throws {
        stubJSON(#"{"id":"f1","name":"Notes"}"#)
        let folder = try await client.createFolder(name: "Notes", parentID: "p1")
        XCTAssertEqual(folder.name, "Notes")
    }

    func testUploadFile() async throws {
        stubJSON(#"{"id":"f1","file_name":"doc.txt","mime_type":"text/plain","total_size":5,"manifest_block_id":"mb1","folder_id":""}"#)
        let data = "hello".data(using: .utf8)!
        let result = try await client.uploadFile(data: data, fileName: "doc.txt", mimeType: "text/plain")
        XCTAssertEqual(result.fileName, "doc.txt")
    }

    func testDownloadFile() async throws {
        let b64 = Data("hello".utf8).base64EncodedString()
        stubJSON(#"{"data_b64":"\#(b64)"}"#)
        let data = try await client.downloadFile(manifestBlockID: "mb1")
        XCTAssertEqual(data.count, 5)
    }

    // ── Workspaces ────────────────────────────────────────────────────────────

    func testListWorkspaces() async throws {
        stubJSON(#"{"workspaces":[{"id":"ws1","name":"Personal","is_default":true}]}"#)
        let workspaces = try await client.listWorkspaces()
        XCTAssertEqual(workspaces.count, 1)
        XCTAssertEqual(workspaces[0].name, "Personal")
    }

    func testCreateWorkspace() async throws {
        stubJSON(#"{"id":"ws2","name":"Work"}"#)
        let ws = try await client.createWorkspace(name: "Work")
        XCTAssertEqual(ws.name, "Work")
    }

    // ── Sync ──────────────────────────────────────────────────────────────────

    func testListSyncPeers() async throws {
        stubJSON(#"{"peers":[{"device_id":"d1","host":"192.168.1.5","port":36352,"seen_at":1}],"last_sync_at":0,"device_id":"d1"}"#)
        let status = try await client.listSyncPeers()
        XCTAssertEqual(status.peers.count, 1)
    }

    func testTriggerSync() async throws {
        stubJSON(#"{"success":true}"#)
        try await client.triggerSync()
    }

    // ── Audit ─────────────────────────────────────────────────────────────────

    func testListAuditEvents() async throws {
        stubJSON(#"{"events":[{"timestamp":1,"level":"INFO","module":"auth","message":"unlock","subject_id":""}]}"#)
        let events = try await client.listAuditEvents()
        XCTAssertEqual(events.count, 1)
        XCTAssertEqual(events[0].level, "INFO")
    }

    // ── Health ────────────────────────────────────────────────────────────────

    func testHealthCheck() async throws {
        stubJSON(#"{"status":"ok","daemon_version":"1.0.0","vault_initialized":true,"vault_unlocked":true}"#)
        let health = try await client.healthCheck()
        XCTAssertEqual(health.status, "ok")
    }

    // ── Tools ─────────────────────────────────────────────────────────────────

    func testGenerateSshKey() async throws {
        stubJSON(#"{"public_key":"ssh-ed25519 AAA","fingerprint":"SHA256:abc","entry_id":"e1"}"#)
        let result = try await client.generateSSHKey(comment: "test", saveToVault: true)
        XCTAssertTrue(result.publicKey.hasPrefix("ssh-ed25519"))
    }

    func testGetRecoveryPhrase() async throws {
        stubJSON(#"{"recovery_phrase":"abandon ability able about above absent"}"#)
        let phrase = try await client.getRecoveryPhrase(password: "master")
        XCTAssertTrue(phrase.hasPrefix("abandon"))
    }

    // ── Error handling ───────────────────────────────────────────────────────

    func testErrorHandlingUnauthorized() async throws {
        stubJSON(#"{"error":"unauthorized"}"#, status: 401)
        do {
            let _ = try await client.listEntries()
            XCTFail("expected error")
        } catch let error as GrimlockerError {
            XCTAssertEqual(error.statusCode, 401)
        }
    }

    func testErrorHandlingInternalError() async throws {
        stubJSON(#"{"error":"internal server error"}"#, status: 500)
        do {
            let _ = try await client.vaultStatus()
            XCTFail("expected error")
        } catch let error as GrimlockerError {
            XCTAssertEqual(error.statusCode, 500)
        }
    }
}

File: grimdb/sdk/csharp/test/GrimlockerClientTests.cs
using System.Net;
using System.Net.Http.Json;
using System.Text.Json;
using Grimlocker.Models;
using Xunit;

namespace Grimlocker.Tests;

public class MockHttpMessageHandler : HttpMessageHandler
{
    private readonly Queue<HttpResponseMessage> _responses;

    public MockHttpMessageHandler(params HttpResponseMessage[] responses)
    {
        _responses = new Queue<HttpResponseMessage>(responses);
    }

    public System.Collections.Generic.IReadOnlyList<HttpRequestMessage> Requests { get; } = new System.Collections.Generic.List<HttpRequestMessage>();

    protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
    {
        ((System.Collections.Generic.List<HttpRequestMessage>)Requests).Add(request);
        var response = _responses.Count > 0 ? _responses.Dequeue() : new HttpResponseMessage(HttpStatusCode.OK)
        {
            Content = JsonContent.Create(new { success = true })
        };
        return Task.FromResult(response);
    }
}

public class GrimlockerClientTests
{
    private static readonly JsonSerializerOptions JsonOpts = new() { PropertyNameCaseInsensitive = true };

    private static MockHttpMessageHandler CreateHandler(params (HttpStatusCode status, object body)[] responses)
    {
        var msgs = responses.Select(r => new HttpResponseMessage(r.status)
        {
            Content = JsonContent.Create(r.body)
        }).ToArray();
        return new MockHttpMessageHandler(msgs);
    }

    [Fact]
    public async Task UnlockVaultAsync_SendsCorrectPayload()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { success = true }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        await client.UnlockVaultAsync("master-password");

        var lastRequest = handler.Requests[0];
        var body = await lastRequest.Content!.ReadFromJsonAsync<System.Text.Json.JsonElement>();
        Assert.Equal("vault.unlock", body.GetProperty("action").GetString());
        Assert.Equal("master-password", body.GetProperty("payload").GetProperty("password").GetString());
    }

    [Fact]
    public async Task LockVaultAsync_SendsLogout()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { success = true }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        await client.LockVaultAsync();

        var body = await handler.Requests[0].Content!.ReadFromJsonAsync<JsonElement>();
        Assert.Equal("vault.logout", body.GetProperty("action").GetString());
    }

    [Fact]
    public async Task VaultStatusAsync_ReturnsStatus()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { initialized = true, unlocked = true, status = "ok" }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var status = await client.VaultStatusAsync();

        Assert.True(status.Initialized);
        Assert.True(status.Unlocked);
    }

    [Fact]
    public async Task ListEntriesAsync_WithoutCategory()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new[] {
            new { id = "e1", title = "Entry 1", category = "PASSWORD", created_at = 1L, updated_at = 2L }
        }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var entries = await client.ListEntriesAsync();

        Assert.Single(entries);
        Assert.Equal("e1", entries[0].Id);
    }

    [Fact]
    public async Task ListEntriesAsync_WithCategory()
    {
        var handler = CreateHandler((HttpStatusCode.OK, Array.Empty<object>()));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        await client.ListEntriesAsync("SSH_KEY");

        var body = await handler.Requests[0].Content!.ReadFromJsonAsync<JsonElement>();
        Assert.Equal("entry.query", body.GetProperty("action").GetString());
    }

    [Fact]
    public async Task GetEntryAsync_ReturnsEntry()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { id = "e1", title = "Entry", category = "PASSWORD" }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var entry = await client.GetEntryAsync("e1");

        Assert.Equal("e1", entry.Id);
    }

    [Fact]
    public async Task GetEntryAsync_ThrowsOnNull()
    {
        var handler = CreateHandler((HttpStatusCode.OK, null!));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        await Assert.ThrowsAsync<GrimlockerException>(() => client.GetEntryAsync("bad"));
    }

    [Fact]
    public async Task CreateEntryAsync_ReturnsCreatedEntry()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { id = "new1", title = "New", category = "PASSWORD" }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var entry = await client.CreateEntryAsync("New", "PASSWORD", new Dictionary<string, string> { ["username"] = "alice" });

        Assert.Equal("new1", entry.Id);
    }

    [Fact]
    public async Task UpdateEntryAsync_SendsUpdate()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { success = true }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        await client.UpdateEntryAsync("e1", new Dictionary<string, string> { ["notes"] = "updated" });

        var body = await handler.Requests[0].Content!.ReadFromJsonAsync<JsonElement>();
        Assert.Equal("entry.update", body.GetProperty("action").GetString());
        Assert.Equal("e1", body.GetProperty("payload").GetProperty("id").GetString());
    }

    [Fact]
    public async Task DeleteEntryAsync_SendsDelete()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { success = true }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        await client.DeleteEntryAsync("e1");

        var body = await handler.Requests[0].Content!.ReadFromJsonAsync<JsonElement>();
        Assert.Equal("entry.delete", body.GetProperty("action").GetString());
        Assert.Equal("e1", body.GetProperty("payload").GetProperty("id").GetString());
    }

    [Fact]
    public async Task SearchEntriesAsync_ReturnsResults()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new[] {
            new { id = "e1", title = "GitHub", category = "PASSWORD" }
        }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var results = await client.SearchEntriesAsync("git");

        Assert.Single(results);
    }

    [Fact]
    public async Task ListPasswordsAsync_ReturnsTypedEntries()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new[] {
            new { id = "p1", title = "GitHub", category = "PASSWORD", fields = new Dictionary<string, string> { ["username"] = "a", ["password"] = "b", ["url"] = "", ["notes"] = "" } }
        }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var passwords = await client.ListPasswordsAsync();

        Assert.Single(passwords);
        Assert.Equal("GitHub", passwords[0].Title);
    }

    [Fact]
    public async Task CreatePasswordAsync_ReturnsId()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { id = "p1", title = "GitHub", category = "PASSWORD" }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var p = new Models.PasswordEntry { Title = "GitHub", Username = "alice", Password = "sec" };
        var id = await client.CreatePasswordAsync(p);

        Assert.Equal("p1", id);
    }

    [Fact]
    public async Task ListSshKeysAsync_ReturnsKeys()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new[] {
            new { id = "sk1", title = "Key", category = "SSH_KEY", fields = new Dictionary<string, string> { ["public_key"] = "pk", ["private_key"] = "", ["username"] = "", ["passphrase"] = "", ["comment"] = "" } }
        }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var keys = await client.ListSshKeysAsync();

        Assert.Single(keys);
    }

    [Fact]
    public async Task CreateSshKeyAsync_ReturnsId()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { id = "sk1", title = "Key", category = "SSH_KEY" }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var k = new Models.SshKeyEntry { Title = "Key", PublicKey = "pk", PrivateKey = "priv" };
        var id = await client.CreateSshKeyAsync(k);

        Assert.Equal("sk1", id);
    }

    [Fact]
    public async Task ListCertificatesAsync_ReturnsCerts()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new[] {
            new { id = "c1", title = "Cert", category = "CERTIFICATE", fields = new Dictionary<string, string> { ["domain"] = "ex.com", ["certificate"] = "crt", ["private_key"] = "key" } }
        }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var certs = await client.ListCertificatesAsync();

        Assert.Single(certs);
    }

    [Fact]
    public async Task CreateCertificateAsync_ReturnsId()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { id = "c1", title = "Cert", category = "CERTIFICATE" }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var c = new Models.CertificateEntry { Title = "Cert", Domain = "ex.com", Certificate = "crt", PrivateKey = "key" };
        var id = await client.CreateCertificateAsync(c);

        Assert.Equal("c1", id);
    }

    [Fact]
    public async Task ListFolderAsync_ReturnsListing()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { folders = Array.Empty<object>(), files = Array.Empty<object>() }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var listing = await client.ListFolderAsync();

        Assert.NotNull(listing);
    }

    [Fact]
    public async Task CreateFolderAsync_ReturnsFolder()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { id = "f1", name = "Notes" }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var folder = await client.CreateFolderAsync("Notes");

        Assert.Equal("Notes", folder.Name);
    }

    [Fact]
    public async Task UploadFileAsync_ReturnsFileEntry()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { id = "f1", file_name = "doc.txt", mime_type = "text/plain", total_size = 12, manifest_block_id = "mb1", folder_id = "" }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var result = await client.UploadFileAsync("hello world"u8.ToArray(), "doc.txt", "text/plain");

        Assert.Equal("doc.txt", result.FileName);
    }

    [Fact]
    public async Task DownloadFileAsync_ReturnsData()
    {
        var b64 = Convert.ToBase64String("hello"u8.ToArray());
        var handler = CreateHandler((HttpStatusCode.OK, new { data_b64 = b64 }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var data = await client.DownloadFileAsync("mb1");

        Assert.Equal(5, data.Length);
    }

    [Fact]
    public async Task ListWorkspacesAsync_ReturnsWorkspaces()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new[] {
            new { id = "ws1", name = "Personal", is_default = true }
        }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var workspaces = await client.ListWorkspacesAsync();

        Assert.Single(workspaces);
        Assert.Equal("Personal", workspaces[0].Name);
    }

    [Fact]
    public async Task CreateWorkspaceAsync_ReturnsWorkspace()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { id = "ws2", name = "Work" }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var ws = await client.CreateWorkspaceAsync("Work");

        Assert.Equal("Work", ws.Name);
    }

    [Fact]
    public async Task ListSyncPeersAsync_ReturnsStatus()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { peers = Array.Empty<object>(), last_sync_at = 0L, device_id = "d1" }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var status = await client.ListSyncPeersAsync();

        Assert.NotNull(status);
    }

    [Fact]
    public async Task TriggerSyncAsync_SendsTrigger()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { success = true }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        await client.TriggerSyncAsync();

        var body = await handler.Requests[0].Content!.ReadFromJsonAsync<JsonElement>();
        Assert.Equal("sync.trigger", body.GetProperty("action").GetString());
    }

    [Fact]
    public async Task ListAuditEventsAsync_ReturnsEvents()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new[] {
            new { timestamp = 1L, level = "INFO", module = "auth", message = "unlock", subject_id = "" }
        }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var events = await client.ListAuditEventsAsync();

        Assert.Single(events);
    }

    [Fact]
    public async Task HealthCheckAsync_ReturnsHealth()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { status = "ok", daemon_version = "1.0.0", vault_initialized = true, vault_unlocked = true }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var health = await client.HealthCheckAsync();

        Assert.Equal("ok", health.Status);
    }

    [Fact]
    public async Task GenerateSshKeyAsync_ReturnsKey()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { public_key = "ssh-ed25519 AAA", fingerprint = "SHA256:abc", entry_id = "e1" }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        var result = await client.GenerateSshKeyAsync();

        Assert.Contains("ssh-ed25519", result.PublicKey);
    }

    [Fact]
    public async Task ErrorHandling_ThrowsOnError()
    {
        var handler = CreateHandler((HttpStatusCode.Unauthorized, new { error = "Invalid token" }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        await Assert.ThrowsAsync<GrimlockerException>(() => client.ListEntriesAsync());
    }

    [Fact]
    public async Task MoveFileAsync_SendsMove()
    {
        var handler = CreateHandler((HttpStatusCode.OK, new { success = true }));
        var client = new TestableGrimlockerClient("http://127.0.0.1:36353", "token", handler);

        await client.MoveFileAsync("mb1", "folder1");

        var body = await handler.Requests[0].Content!.ReadFromJsonAsync<JsonElement>();
        Assert.Equal("file.move", body.GetProperty("action").GetString());
    }

    /// <summary>
    /// Testable client that accepts a custom HttpMessageHandler.
    /// </summary>
    private class TestableGrimlockerClient : GrimlockerClient
    {
        public TestableGrimlockerClient(string baseUrl, string token, HttpMessageHandler handler)
            : base(baseUrl, token)
        {
            // Replace the HttpClient's handler
            var httpField = typeof(GrimlockerClient)
                .GetField("_http", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance)!;
            var newHttp = new HttpClient(handler);
            newHttp.DefaultRequestHeaders.Add("X-Grimlocker-Token", token);
            httpField.SetValue(this, newHttp);
        }
    }
}
